core/hybrisadaptor.cpp

```diff
@@ -937,6 +1100,77 @@ void HybrisManager::pollEventsCallback(
     manager->pollEvents();
 }
 
+void *HybrisManager::binderEventReaderThread(void *aptr)
+{
+    HybrisManager *manager = static_cast<HybrisManager *>(aptr);
+    static const size_t maxNumEvents = 16;
+    sensors_event_t buffer[maxNumEvents];
+
+    /* Async cancellation, but disabled */
+    pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, 0);
+    pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, 0);
+    /* Leave INT/TERM signal processing up to the main thread */
+    sigset_t ss;
+    sigemptyset(&ss);
+    sigaddset(&ss, SIGINT);
+    sigaddset(&ss, SIGTERM);
+    pthread_sigmask(SIG_BLOCK, &ss, 0);
+    /* Loop until explicitly canceled */
+    for (;;) {
+        /* Async cancellation point */
+        pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, 0);
+        size_t numberOfEvents = gbinder_fmq_available_to_read(manager->m_eventQueue);
+        pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, 0);
+        /* Rate limit in poll() error situations */
+
+        if (numberOfEvents <= 0) {
+            guint32 state = 0;
+            gint32 ret = gbinder_fmq_wait_timeout(manager->m_eventQueue,
+                EVENT_QUEUE_FLAG_READ_AND_PROCESS, &state, 1000 /* 1 sec */);
+
+            if (ret < 0) {
+                if (ret != -ETIMEDOUT && ret != -EAGAIN) {
+                    sensordLogW() << "Waiting for events failed" << strerror(-ret);
+                }
+                continue;
+            }
+            numberOfEvents = gbinder_fmq_available_to_read(manager->m_eventQueue);
+        }
+        size_t eventsToRead = std::min(numberOfEvents, maxNumEvents);
+        if (eventsToRead == 0) {
+            continue;
+        }
+        if (gbinder_fmq_read(manager->m_eventQueue, buffer, eventsToRead)) {
+            gbinder_fmq_wake(manager->m_eventQueue, EVENT_QUEUE_FLAG_EVENTS_READ);
+        } else {
+            sensordLogW() << "Reading events failed";
+            continue;
+        }
+
+        // Process received events
+        int blockSuspend = 0;
+        bool errorInInput = false;
+        manager->processEvents(buffer, numberOfEvents, blockSuspend, errorInInput);
+
+        // Suspend proof sensor reporting that could occur in display off
+        if (blockSuspend) {
+            ObtainTemporaryWakeLock();
+            if (gbinder_fmq_write(manager->m_wakeLockQueue, &blockSuspend, 1)) {
```

Comment:
What is happening here? I'm guessing this is whatever it is that made blockSuspend type change from bool to int necessary.

Reply:
In the documentation for Sensors 2 API (https://android.googlesource.com/platform/hardware/interfaces/+/refs/heads/master/sensors/2.1/ISensors.hal#62) it said that the HAL must be notified of how many wake up sensor events have been processed so the HAL knows if it can release any wake locks it has acquired.

Reply:
ok, I see. But now I'm wondering if the counting is done correctly. What hal needs to be told (presumably: number of events form sensors with SENSOR_FLAG_WAKE_UP?) is not exactly what blockSuspend was / is (number of proximity sensor events, whether they are from wake up sensor or not) -> it works for our purposes (p-sensor data xfer is still suspend proofed), but hal might get incorrect numbers (with some chance of perpetual suspend blocking).

In all likelihood proximity sensor is of wake up kind, so it might be enough to revise the conditions for incrementing blockSuspend value from "is p-sensor" to "is wake-up sensor". [Also, the naming is a bit off now, something like wakeupEventCount might be more descriptive]

Reply:
If you look at the changes I made to HybrisManager::initManager() you can see I now pick the wake-up variant of sensor which are supposed to be wake-up sensors by default according to Android sources. I also changed the behavior of HybrisManager::processEvents to assume wake-up flag is correct in case of binder support, documentation says it should be always correctly set for the hidl API sensors. Previously we always picked the first sensor of a type but I noticed that on Xperia 10 III the non-wake-up variant was listed first.